Add subscriptions support (#13738)
DataDog/datadog-agent: docs/components/registrations.md

In simple cases, that occurs by method calls.
But in many cases, a single component needs to communicate with a number of other components that all share some characteristics.
For example, the `comp/core/health` component monitors the health of many other components, and `comp/workloadmeta/scheduler` provides workload events to an arbitrary number of subscribers.

The convention in the Agent codebase is to use [value groups](./fx.md#value-groups) to accomplish this.
The _collecting_ component requires a slice of some _collected type_, and the _providing_ components provide values of that type.
Consider a simple case of an HTTP server component to which endpoints can be attached.
The server is the collecting component, requiring a slice of type `[]*endpoint`, where `*endpoint` is the collected type.
Providing components provide values of type `*endpoint`.

The convention is to "wrap" the collected type in a struct type which embeds `fx.Out` and has tag `group:"pkgname"`, where `pkgname` is the short package name (Fx requires a group name, and this is as good as any).
This helps providing components avoid the common mistake of omitting the tag.

> The collected type can be exported, such as if it has useful methods for providing components to call, but can also be unexported.

Reviewer: ```suggestion
The collected type can be exported, for example, if it has useful methods for providing components to call, but does not need to be exported.
```

Author: "unexported" is the term used for lower-case item names in Go, but this sentence is pretty confusing anyway.  Let me see what I can do :)

Collecting components require `subcriptions.Publisher[coll.Message]`, from which they can obtain a `subscriptions.Transmitter[coll.Message]`.

```go
// --- announcer/component.go ---

// ...
// To subscribe to these announcements, provide a subscriptions.Subscription[announcer.Announcement].
// ...
package announcer
```

```go
// --- announcer/announcer.go ---

func newAnnouncer(pub subscriptions.Publisher[Anouncement]) Component {
    return &announcer{announcementTx: pub.Transmitter()}  // (get a Transmitter from the Publisher)
}

// .. later send messages with 
func (ann *announcer) announce(a announcement) {
    ann.eventTx.Notify(a)
}
```

```go
// --- listener/listener.go ---

func newListener() (Component, subscriptions.Subscription[announcer.Announcement]) {
    sub := subscriptions.NewSubscription[Event]() // create a subscription
    return &listener{eventRx: sub.Receiver}, sub  // capture the receiver, and return the subscription
}

// .. later receive messages (usually in an actor's main loop)
func (l *listener) run() {
    loop {
        select {
        case a := <- l.eventRx.Chan():
            ...
        }
    }
}
```

> If a receiving component decides it does not want to subscribe after all (such as, if it is not started), it can return the zero value, `subscriptions.Subscription[Event]{}`, from its constructor.
> If a component returns a non-nil subscriber, though, then it _must_ consume messages from the receiver or risk blocking the transmitter.

Reviewer: ```suggestion
If a receiving component does not subscribe (for example, if it is not started), it can return the zero value, `subscriptions.Subscription[Event]{}`, from its constructor.
If a component returns a non-nil subscriber, it _must_ consume messages from the receiver or risk blocking the transmitter.
```

# Registrations

Components generally need to talk to one another!
In simple cases, that occurs by method calls.
But in many cases, a single component needs to communicate with a number of other components that all share some characteristics.
For example, the `comp/core/health` component monitors the health of many other components, and `comp/workloadmeta/scheduler` provides workload events to an arbitrary number of subscribers.

> The convention in the Agent codebase is to use [value groups](./fx.md#value-groups) to accomplish this.

Reviewer 2: In this case, I think this is fine.

 * Fx will instantiate _all_ declared providing components before the collecting component, regardless of whether their `Component` type is required.
   This may lead to components being instantiated in unexpected circumstances.

## Subscriptions

Subscriptions are a common form of registration, and have support in the `pkg/util/subscriptions` package.

To implement a subscription, the collecting component defines a message type.
Providing components provide `subscriptions.Subscription[coll.Message]`, from which they can obtain a `subscriptions.Receiver[coll.Message]`.
Collecting components require `subcriptions.Publisher[coll.Message]`, from which they can obtain a `subscriptions.Transmitter[coll.Message]`.

```go
// --- announcer/component.go ---

// ...
// To subscribe to these announcements, provide a subscriptions.Subscription[announcer.Announcement].
// ...
package announcer
```

```go
// --- announcer/announcer.go ---

func newAnnouncer(pub subscriptions.Publisher[Anouncement]) Component {
    return &announcer{announcementTx: pub.Transmitter()}  // (get a Transmitter from the Publisher)
}

// .. later send messages with 

> func (ann *announcer) announce(a announcement) {

Reviewer 2: I think it is important to write somewhere in a very explicit way (Maybe that is already the case)  that when using `subscriptions.Subscription[SOMETHING]` SOMETHING has to be a new type and cannot be a raw `string` for example.
